File: frontend/src/api/providers.ts
import api from './client'

export type ProviderType =
  | 'ollama' | 'openai' | 'anthropic' | 'gemini'
  | 'openai_compatible' | 'openrouter' | 'opencode_zen' | 'nvidia'
export type ProviderEnvironment = 'local' | 'cloud' | 'custom'

export interface ProviderResponse {
  id: string
  name: string
  provider_type: ProviderType
  environment: ProviderEnvironment
  base_url: string | null
  /** Legacy pinned model — "" when the provider relies on auto-discovery */
  model_name: string
  has_api_key: boolean       // true = key stored; key itself is NEVER returned
  /** Masked hint, e.g. "••••••••abcd" — at most the last 4 characters */
  api_key_masked: string | null
  enabled: boolean
  supports_streaming: boolean
  supports_embeddings: boolean
  description: string | null
  /** Selectable (available + enabled) discovered models */
  model_count: number
  created_at: string
  updated_at: string
}

/** Provider preset shown in the Add-Provider picker */
export interface ProviderPreset {
  id: string
  label: string
  description: string
  provider_type: ProviderType
  environment: ProviderEnvironment
  default_base_url: string
  requires_api_key: boolean
  api_key_hint: string | null
  api_key_url: string | null
  supports_discovery: boolean
  adapter: string
}

/** A persisted discovered model belonging to a provider */
export interface ModelRecord {
  id: string                 // record id (used for enable/disable)
  provider_id: string
  model_id: string           // provider-native name, e.g. "llama3.2:3b"
  display_name: string | null
  family: string | null
  parameter_size: string | null
  quantization: string | null
  size_bytes: number | null
  modified_at: string | null
  context_length: number | null
  status: 'available' | 'unavailable'
  enabled: boolean
}

export interface ProviderCreate {
  name: string
  provider_type: ProviderType
  environment?: ProviderEnvironment
  base_url?: string
  /** Optional legacy field — providers no longer require a manually typed model */
  model_name?: string
  api_key?: string            // write-only
  enabled?: boolean
  supports_streaming?: boolean
  supports_embeddings?: boolean
  description?: string
}

export interface ProviderUpdate {
  name?: string
  environment?: ProviderEnvironment
  base_url?: string
  model_name?: string
  api_key?: string            // write-only; null = don't change
  enabled?: boolean
  supports_streaming?: boolean
  supports_embeddings?: boolean
  description?: string
}

export interface ProviderTestResult {
  success: boolean
  latency_ms: number | null
  provider: string
  model: string
  error: string | null
  /** Number of models discovered from the provider during the test (if listing succeeded) */
  models_found: number | null
}

/** A model available from a connected provider — only fields the provider returns are set */
export interface DiscoveredModel {
  name: string
  display_name: string
  family: string | null
  parameter_size: string | null
  quantization: string | null
  size_bytes: number | null
  modified_at: string | null
  context_length: number | null
  vision_capable: boolean | null
  embedding_capable: boolean | null
  status: string
}

export const PROVIDER_LABELS: Record<ProviderType, string> = {
  ollama: 'Ollama',
  openai: 'OpenAI',
  anthropic: 'Anthropic',
  gemini: 'Google Gemini',
  openai_compatible: 'OpenAI-Compatible',
  openrouter: 'OpenRouter',
  opencode_zen: 'OpenCode Zen',
  nvidia: 'NVIDIA NIM',
}

export const ENV_LABELS: Record<ProviderEnvironment, string> = {
  local: '🖥 Local',
  cloud: '☁️ Cloud',
  custom: '⚙️ Custom',
}

export const providersApi = {
  list: () => api.get<ProviderResponse[]>('/models/providers/').then(r => r.data),
  get: (id: string) => api.get<ProviderResponse>(`/models/providers/${id}`).then(r => r.data),
  presets: () => api.get<ProviderPreset[]>('/models/providers/presets').then(r => r.data),
  create: (data: ProviderCreate) =>
    api.post<ProviderResponse>('/models/providers/', data).then(r => r.data),
  update: (id: string, data: ProviderUpdate) =>
    api.put<ProviderResponse>(`/models/providers/${id}`, data).then(r => r.data),
  delete: (id: string) => api.delete(`/models/providers/${id}`),
  test: (id: string) =>
    api.post<ProviderTestResult>(`/models/providers/${id}/test`).then(r => r.data),
  /** Live discovery + upsert of the persisted catalog */
  refreshModels: (id: string) =>
    api.post<ModelRecord[]>(`/models/providers/${id}/refresh`).then(r => r.data),
  /** Persisted catalog — no live call */
  getModels: (id: string) =>
    api.get<ModelRecord[]>(`/models/providers/${id}/models`).then(r => r.data),
  setModelEnabled: (providerId: string, recordId: string, enabled: boolean) =>
    api.patch<ModelRecord>(`/models/providers/${providerId}/models/${recordId}`,
      { enabled }).then(r => r.data),
}

/** Per-user preferred chat model */
export interface ModelPreference {
  provider_id: string | null
  model_name: string | null
}

export const prefsApi = {
  get: () => api.get<ModelPreference>('/models/preferences').then(r => r.data),
  set: (provider_id: string | null, model_name: string) =>
    api.put<ModelPreference>('/models/preferences',
      { provider_id, model_name }).then(r => r.data),
}

File: backend/routers/providers.py
"""
LLM Provider management router.

Security:
- GET endpoints NEVER return api_key (only has_api_key + api_key_masked hint)
- Create/Update/Delete/Test/enable-model are admin-only
- List/Get/model-catalog/presets/preferences are available to any authenticated user
- Model refresh queries the live provider via its adapter (read-only)
"""
from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.ext.asyncio import AsyncSession

from database import get_db
from dependencies import get_current_user, require_role, get_client_ip
from models.provider import PROVIDER_PRESETS
from models.user import User
from schemas.provider import (
    DiscoveredModel, ModelEnabledUpdate, ModelRecord,
    ProviderCreate, ProviderPreset, ProviderResponse,
    ProviderTestResult, ProviderUpdate,
)
from services.provider_service import ProviderService

router = APIRouter(tags=["providers"])


def _get_svc(db: AsyncSession = Depends(get_db)) -> ProviderService:
    return ProviderService(db)


# ── Presets (any authenticated user) ────────────────────────────
# NOTE: declared before /{provider_id} routes so "presets" is not treated as an id.
@router.get("/presets", response_model=list[ProviderPreset])
async def list_presets(_user=Depends(get_current_user)):
    """Provider presets for the Add-Provider picker."""
    return [ProviderPreset(**p) for p in PROVIDER_PRESETS]


# ── List (any authenticated user) ──────────────────────────────
@router.get("/", response_model=list[ProviderResponse])
async def list_providers(
    _user=Depends(get_current_user),
    svc: ProviderService = Depends(_get_svc),
):
    """Return all providers - api_key is never included."""
    return await svc.list_providers()


# ── Get (any authenticated user) ───────────────────────────────
@router.get("/{provider_id}", response_model=ProviderResponse)
async def get_provider(
    provider_id: str,
    _user=Depends(get_current_user),
    svc: ProviderService = Depends(_get_svc),
):
    return await svc.get(provider_id)


# ── Cached model catalog (any authenticated user) ──────────────
@router.get("/{provider_id}/models", response_model=list[ModelRecord])
async def get_provider_models(
    provider_id: str,
    _user=Depends(get_current_user),
    svc: ProviderService = Depends(_get_svc),
):
    """Return the persisted model catalog for a provider (no live call)."""
    return await svc.get_cached_models(provider_id)


# ── Refresh models from the live provider (any authenticated user) ──
@router.post("/{provider_id}/refresh", response_model=list[ModelRecord])
async def refresh_provider_models(
    provider_id: str,
    _user=Depends(get_current_user),
    svc: ProviderService = Depends(_get_svc),
):
    """
    Discover models live from the provider and upsert the local catalog.
    Newly installed models appear; removed ones are marked unavailable.
    """
    try:
        return await svc.refresh_models(provider_id)
    except HTTPException:
        raise


# ── Enable/disable a discovered model (admin only) ─────────────
@router.patch(
    "/{provider_id}/models/{model_record_id}",
    response_model=ModelRecord,
)
async def set_model_enabled(
    provider_id: str,
    model_record_id: str,
    data: ModelEnabledUpdate,
    request: Request,
    admin: User = Depends(require_role("admin")),
    svc: ProviderService = Depends(_get_svc),
):
    """Enable or disable a discovered model for selection in Chat."""
    return await svc.set_model_enabled(
        provider_id, model_record_id, data.enabled, user_id=admin.id,
        client_ip=get_client_ip(request)
    )


# ── Create (admin only) ────────────────────────────────────────
@router.post("/", response_model=ProviderResponse, status_code=201)
async def create_provider(
    data: ProviderCreate,
    request: Request,
    admin: User = Depends(require_role("admin")),
    svc: ProviderService = Depends(_get_svc),
):
    return await svc.create(data, user_id=admin.id, client_ip=get_client_ip(request))


# ── Update (admin only) ────────────────────────────────────────
@router.put("/{provider_id}", response_model=ProviderResponse)
async def update_provider(
    provider_id: str,
    data: ProviderUpdate,
    request: Request,
    admin: User = Depends(require_role("admin")),
    svc: ProviderService = Depends(_get_svc),
):
    return await svc.update(provider_id, data, user_id=admin.id, client_ip=get_client_ip(request))


# ── Delete (admin only) ────────────────────────────────────────
@router.delete("/{provider_id}", status_code=204)
async def delete_provider(
    provider_id: str,
    request: Request,
    admin: User = Depends(require_role("admin")),
    svc: ProviderService = Depends(_get_svc),
):
    await svc.delete(provider_id, user_id=admin.id, client_ip=get_client_ip(request))


# ── Test connection (admin only) ───────────────────────────────
@router.post("/{provider_id}/test", response_model=ProviderTestResult)
async def test_provider(
    provider_id: str,
    _admin=Depends(require_role("admin")),
    svc: ProviderService = Depends(_get_svc),
):
    """
    Test connectivity to a provider.
    Returns success/latency/error/models_found — NEVER returns the api_key.
    """
    return await svc.test_connection(provider_id)
